lxd: Update logic for project config patch

boltmark opened the pull request:
This PR provides an update to the logic for patching project configurations. Previously, there was an overwrite of the existing project config such that it was updated to consist solely of the data in the patch. With this update we maintain existing key/value pairs, and only overwrite those also present in the patch.

Closes #13632.

tomponline commented on lxd/api_project.go:
```diff
@@ -599,16 +599,18 @@ func projectPatch(d *Daemon, r *http.Request) response.Response {
 			}
 		}
+	} else {
+		logger.Error("Failed retrieving config map from request", logger.Ctx{"err": err})
```
Does this fire if no input config is supplied by the request? 

I'm wondering why the current version has:

```go
config, err := reqRaw.GetMap("config")
	if err != nil {
		req.Config = project.Config
```

As I don't think we should be logging an error if a valid request with no config comes in for the patch request, we should just not updating anything in the config part.

boltmark replied:
Good point -- with the incoming changes I do see the error logged if no input config is supplied. I agree that it's not a good idea to log an error here since the input is technically valid. Will fix this.

tomponline commented:
@boltmark Please can you add a test for updating projects via PATCH, see `test_certificate_edit` function in the tests dir for an example of patch tests.

```
my_curl -X PATCH -d "{\"certificate\":\"$(sed ':a;N;$!ba;s/\n/\\n/g' "${LXD_CONF}/client-new.crt")\"}" "https://${LXD_ADDR}/1.0/certificates/${FINGERPRINT}"
```

You can add it to the projects.sh file in the `test_projects_crud` function.

boltmark commented:
@tomponline Thanks for the comments. I have added testing for updating project configs via PATCH in `projects.sh` under `test_projects_crud`. I have also removed the error logging as per the above comment. This is ready for another review!

tomponline reviewed (approved):
Thanks!
